crates/uv-installer/src/satisfies.rs

```diff
@@ -18,53 +17,139 @@ impl RequirementSatisfaction {
-        match version_or_url {
-            // Accept any version of the package.
-            None => return Ok(Self::Satisfied),
-
+        match source {
             // If the requirement comes from a registry, check by name.
-            Some(VersionOrUrlRef::VersionSpecifier(version_specifier)) => {
-                if version_specifier.contains(distribution.version()) {
+            RequirementSource::Registry { specifier, .. } => {
+                if specifier.contains(distribution.version()) {
                     return Ok(Self::Satisfied);
                 }
+                Ok(Self::Mismatch)
             }
+            RequirementSource::Url {
+                // We use the location since `direct_url.json` also stores this URL, e.g.
+                // `pip install git+https://github.com/tqdm/tqdm@cc372d09dcd5a5eabdc6ed4cf365bdb0be004d44#subdirectory=.`
+                // records `"url": "https://github.com/tqdm/tqdm"` in `direct_url.json`.
+                location: requested_url,
+                subdirectory: requested_subdirectory,
+                url: _,
+            } => {
+                let InstalledDist::Url(InstalledDirectUrlDist {
+                    direct_url,
+                    editable,
+                    ..
+                }) = &distribution
+                else {
+                    return Ok(Self::Mismatch);
+                };
+                let DirectUrl::ArchiveUrl {
+                    url: installed_url,
+                    archive_info: _,
+                    subdirectory: installed_subdirectory,
+                } = direct_url.as_ref()
+                else {
+                    return Ok(Self::Mismatch);
+                };
 
-            // If the requirement comes from a direct URL, check by URL.
-            Some(VersionOrUrlRef::Url(url)) => {
-                if let InstalledDist::Url(installed) = &distribution {
-                    if !installed.editable && &installed.url == url.raw() {
-                        // If the requirement came from a local path, check freshness.
-                        return if let Some(archive) = (url.scheme() == "file")
-                            .then(|| url.to_file_path().ok())
-                            .flatten()
-                        {
-                            if ArchiveTimestamp::up_to_date_with(
-                                &archive,
-                                ArchiveTarget::Install(distribution),
-                            )? {
-                                return Ok(Self::Satisfied);
-                            }
-                            Ok(Self::OutOfDate)
-                        } else {
-                            // Otherwise, assume the requirement is up-to-date.
-                            Ok(Self::Satisfied)
-                        };
+                if *editable {
+                    return Ok(Self::Mismatch);
+                }
+
+                if &requested_url.to_string() != installed_url
```

Comment:
This seems off. Shouldn't we instead parse the URL, and use `CanonicalUrl` to compare these? Why are we comparing as strings?

Reply:
(Fixed in https://github.com/astral-sh/uv/pull/3373)